/*
 * Copyright (c) 2010  David Sorokin <david.sorokin@gmail.com>
 * Copyright (c) 2010  rampelstinskin@gmail.com
 * All rights reserved.
 *
 * Redistribution and use in source and binary forms, with or without
 * modification, are permitted provided that the following conditions
 * are met:
 *    1. Redistributions of source code must retain the above copyright
 *       notice, this list of conditions and the following disclaimer.
 *    2. Redistributions in binary form must reproduce the above copyright
 *       notice, this list of conditions and the following disclaimer in the
 *       documentation and/or other materials provided with the distribution.
 * 
 * THIS SOFTWARE IS PROVIDED BY THE AUTHOR ``AS IS'' AND ANY EXPRESS OR
 * IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE IMPLIED WARRANTIES
 * OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE DISCLAIMED. IN
 * NO EVENT SHALL THE AUTHOR BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL,
 * SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED
 * TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR
 * PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF
 * LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING
 * NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF THIS
 * SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
 */

using Nemerle;
using Nemerle.Utility;
using Nemerle.Collections;

using System;
using System.Collections.Generic;
using System.Linq;

namespace Nemerle.ComputationExpressions.Internal
{
  /// <summary>
  /// Represents an enumeration that imitates the try-catch block.
  /// </summary>
  [Record]
  public class TryCatchEnumerable[T] : IEnumerable[T]
  {
    private body : IEnumerable[T];
    private cont : Exception -> IEnumerable[T];
    
    public GetEnumerator () : IEnumerator[T]
    {
      try
      {
        SubEnumerator (body.GetEnumerator (), cont)
      }
      catch
      {
        | e is Exception =>
          cont (e).GetEnumerator ()
      }
    }
    
    private class SubEnumerator : IEnumerator[T]
    {
      private mutable ts : IEnumerator[T];
      private mutable cs : IEnumerator[T];
      private cont : Exception -> IEnumerable[T];
      
      public this (ts : IEnumerator[T], cont : Exception -> IEnumerable[T])
      {
        this.ts = ts;
        this.cs = null;
        this.cont = cont;
      }
      
      public Dispose () : void
      {
        when (ts != null)
          ts.Dispose ();
        
        when (cs != null)
          cs.Dispose ();
      }
      
      public Reset () : void
      {
        throw NotSupportedException ()
      }
      
      public MoveNext () : bool
      {
        if (cs != null)
          cs.MoveNext ()
        else
          try
          {
            ts.MoveNext ()
          }
          catch
          {
            | e is Exception =>
            
              ts.Dispose ();
              ts = null;
            
              cs = cont (e).GetEnumerator ();
              cs.MoveNext ();
          }
      }

      public Current : T
      {
        get { if (ts != null) ts.Current else cs.Current }
      }
    }
  }
}
